server/service/microsoft_mdm.go

```diff
@@ -1181,17 +1187,187 @@ func (svc *Service) GetMDMWindowsTOSContent(ctx context.Context, redirectUri str
+	}
+
+	// TODO - Security checks that need to be implemented
+	// - TLS based auth
+	// - Device auth based on Source/LocURI DeviceID information (this should be present on Enrollment DB)
+
+	return nil
+}
+
+// processIncomingProtocolCommands will process the incoming command
+func (svc *Service) processIncomingProtocolCommands(messageID string, deviceID string, cmd mdm_types.ProtoCmdOperation) (*fleet.SyncMLCmd, error) {
+	if svc == nil {
+		return nil, fmt.Errorf("service was not initialized")
+	}
+
+	// TODO:
+	// - Handle Security Alerts
+	// - Handle Results/Responses
+
+	// We return a status 200 for all the operations
+	return NewSyncMLCmdStatus(messageID, cmd.Cmd.CmdID, string(cmd.Verb), mdm.CmdStatusCode200), nil
+}
+
+// processIncomingMDMCmds process the incoming message from the device
+// It will return the list of operations that need to be sent to the device
+func (svc *Service) processIncomingMDMCmds(deviceID string, reqMsg *fleet.SyncML) ([]*fleet.SyncMLCmd, error) {
+	if svc == nil {
+		return nil, fmt.Errorf("service was not initialized")
+	}
+
+	var responseCmds []*fleet.SyncMLCmd
+
+	// Get the incoming MessageID
+	reqMessageID, err := reqMsg.GetMessageID()
+	if err != nil {
+		return nil, fmt.Errorf("processing incoming msg %v", err)
+	}
+
+	// Acknowledge the message header
+	// msgref is always 0 for the header
+	if err = reqMsg.IsValidHeader(); err == nil {
+		ackMsg := NewSyncMLCmdStatus(reqMessageID, "0", mdm.SyncMLHdrName, mdm.CmdStatusCode200)
+		responseCmds = append(responseCmds, ackMsg)
+	}
+
+	// Now we need to check for any operations that need to be processed
+	protoCMDs := reqMsg.GetOrderedCmds()
+
+	// Iterate over the operations and process them
+	for _, protoCMD := range protoCMDs {
+		protoCmd, err := svc.processIncomingProtocolCommands(reqMessageID, deviceID, protoCMD)
+		if err != nil {
+			return nil, fmt.Errorf("processing incoming msg %v", err)
+		}
+
+		// Append the operations to the response
+		if (protoCmd != nil) && (protoCmd.IsValid()) {
+			responseCmds = append(responseCmds, protoCmd)
+		}
+	}
+
+	return responseCmds, nil
+}
+
+// getPendingMDMCmds returns the list of pending MDM commands for the device
+func (svc *Service) getPendingMDMCmds(ctx context.Context, deviceID string) ([]*mdm_types.SyncMLCmd, error) {
+	if svc == nil {
+		return nil, fmt.Errorf("service was not initialized")
+	}
+
+	// Getting the list of pending commands
+	pendingCmds, err := svc.ds.MDMWindowsListPendingCommands(ctx, deviceID)
+	if err != nil {
+		return nil, fmt.Errorf("getting incoming cmds %v", err)
+	}
+
+	// Converting the pending commands to its target SyncML types
+	var cmds []*mdm_types.SyncMLCmd
+	for _, pendingCmd := range pendingCmds {
+		newCmd, err := NewTypedSyncMLCmd(mdm_types.SyncMLDataType(pendingCmd.DataType), pendingCmd.CmdVerb, pendingCmd.SettingURI, pendingCmd.SettingValue)
+		if (err != nil) || (newCmd == nil) {
+			continue
```

Comment:
well, I'm not sure. As this is called to create the SyncML message to send over the wire, I was concerned that if we ran into an error, it might end up spamming the error log, but we would miss catching the error condition. I'm going to log an error now